feat: add tekton-tasks-operator

codingben opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Add kubevirt/tekton-tasks-operator here to have fewer operators and reduce maintenance efforts. 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add tekton-tasks-operator
```

ksimon1 commented on internal/common/labels.go:
```diff
@@ -13,6 +13,10 @@ const (
+
+	AppComponentTektonTasks     AppComponent = "tektonTasks"
+	AppComponentTektonPipelines AppComponent = "tektonPipelines"
+	AppKubernetesManagedByValue              = "tekton-tasks-operator"
```
this should change to ssp operator

ksimon1 commented on internal/environment/environment.go:
```diff
@@ -0,0 +1,131 @@
+	return duration, nil
+}
+
+func GetOperatorVersion() string {
```
This fn already exists in https://github.com/kubevirt/ssp-operator/blob/master/internal/common/environment.go

ksimon1 commented on internal/environment/environment.go:
```diff
@@ -0,0 +1,131 @@
+	return EnvOrDefault(OperatorVersionKey, defaultOperatorVersion)
+}
+
+func EnvOrDefault(envName string, defVal string) string {
```
https://github.com/kubevirt/ssp-operator/blob/master/internal/common/environment.go#L26

ksimon1 commented on internal/environment/environment.go:
```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	OperatorVersionKey        = "OPERATOR_VERSION"
```
https://github.com/kubevirt/ssp-operator/blob/master/internal/common/environment.go#L19

codingben commented:
@ksimon1 I'd prefer to keep everything in one commit until I'll have no errors and CI passed, to not deal with updating multiple commits. Once it's ready to be merged, I'll split PR into commits. Sounds good?

ksimon1 commented:
Fine by me

akrejcir commented:
Is the `TektonTasks` API going to be a separate CRD, that will be watched by the SSP operator? If yes, then it needs a separate controller.

ksimon1 commented:
> Is the `TektonTasks` API going to be a separate CRD, that will be watched by the SSP operator? If yes, then it needs a separate controller.

No, we will insert it inside already existing CRD

akrejcir commented:
Ok, in that case, @codingben, please move all the API files into `v1beta1` directory, so that the version is the same. 
You can also remove the `TektonTasks`, `TektonTasksList` and `TektonTasksStatus` structs, and keep only `TektonTasksSpec`.

akrejcir commented:
Do we need to keep the API definitions of `TektonTasks` for backward compatibility?

lyarwood commented:
/cc

codingben commented:
FYI, I've rebased this PR on latest tekton-tasks-operator changes [1] and rebased on latest master changes [2]. Also `make lint` passed now.

[1] https://github.com/kubevirt/tekton-tasks-operator/commit/be9190ec25d1dc188e93040f6f193437c11b85d8
[2] https://github.com/kubevirt/ssp-operator/commit/d12622e82bdc397d47922dca0c4e7a9f07d0fe6c

akrejcir reviewed (commented):
I looked at this only partially.

Before you post this for proper review, can you split it to multiple commits, so that the first commit will be only a simple copy-paste of code from TTO repo, and the subsequent commits will contain other changes. Then it will be easier to see what has changed from TTO.

codingben commented:
Hi @akrejcir, a quick question, you've mentioned: `so that the first commit will be only a simple copy-paste of code from TTO repo`.

There is no simple copy-paste because it requires some code changes, in order to work in ssp-operator (especially functional tests). What I want to do is:

- feat: update SSP api
- feat: add tekton-pipelines operand
- feat: add tekton-tasks operand
- test: add tekton functional tests
- other commits

Sounds good?

akrejcir commented:
I've checked the TTO code, and it is mostly compatible with SSP codebase, so copy-paste should be possible, even if the commit does not compile by itself. I think it will make review easier, if I can see what code was copied from TTO and what code is new.

I would propose to split it to commits like this:
1. Copy code related only to `tekton-tasks` operand from TTO to SSP. This would also include functional tests and any common code used by the operand, like `tekton-bundle`.
2. Add a new field to the `SSPSpec` in the `api` package, and modify the `tekton-tasks` code so that it compiles, and passes unittests. Ideally it should pass functests too, but that would take a long time to run tests.
3. Copy code related to `tekton-pipelines` operand from TTO to SSP, including functional tests.
4. Add new filed to the `SSPSpec`, and fix the code, so that it compiles and passes unit tests.
5. Copy anything that's left, like `csv-generator` and CI scripts.
6. Modify them if needed, so that they compile and work correctly.
7. Any other refactoring to make the code simpler and easier to read.

It will be some work, but it will be easier to review and spot potential bugs.

codingben commented on automation/common/versions.sh:
```diff
@@ -15,7 +15,7 @@ function latest_patch_version() {
     local versions_in_page
     versions_in_page=$(
-      curl --fail -s "https://api.github.com/repos/kubevirt/${repo}/releases?per_page=100&page=${page}" |
+      curl --fail -s "https://api.github.com/repos/${org}/${repo}/releases?per_page=100&page=${page}" |
```
It's defined when you call to the function. Looks like `latest_patch_version` isn't called.

codingben replied:
Done. Sorry for the confusion.

codingben commented:
Hi @ksimon1 @akrejcir @lyarwood @0xFelix, it's ready for review. I've spitted this PR to multiple commits. Each commit is like small PR.

codingben commented:
The code in this PR was compiled and works, including unit tests (and `make container-build`). What wasn't tested is the new functional tests.

akrejcir reviewed (commented):
I've reviewed the first 2 commits. I wrote a lot of comments on the first commit, but they apply to the second one, because github does not allow me to comment on lines that are not changed by a commit.

Comments marked as `code style` are minor changes, that can be done in a new commit at the end of this PR.

0xFelix commented on controllers/ssp_controller.go:
```diff
@@ -452,6 +454,26 @@ func listExistingCRDKinds(sspRequest *common.Request) []string {
 	return foundKinds
 }
 
+func isMissingCRDs(crdWatch *crd_watch.CrdWatch, requiredCRDs []string) bool {
```
Can moved to be an extension of `CrdWatch`? Looks to me like it belongs there.

0xFelix reviewed (commented):
I only managed to do a rather limited review given the size of the PR, will continue with reviewing next week.

ksimon1 commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+		}
+
+		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
+			namespace := request.Instance.Namespace
```
yep, namespace should be taken from .spec.Pipelines.Namespace (only if possible)

akrejcir replied:
yes

codingben replied:
Done.

codingben commented on controllers/setup.go:
```diff
@@ -56,13 +58,23 @@ func setupManager(ctx context.Context, cancel context.CancelFunc, mgr controller
 	}
 
+	reader := mgr.GetAPIReader()
+	tektonPipelinesBundle, err := tekton_bundle.ReadPipelineBundle(reader, ctx)
```
Yep, done.

codingben commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -0,0 +1,179 @@
+	ConfigMaps      []v1.ConfigMap
+}
+
+func ReadPipelineBundle(cl client.Reader, ctx context.Context) (*Bundle, error) {
```
Done.

codingben commented on config/samples/ssp_v1beta1_ssp.yaml:
```diff
@@ -8,3 +8,7 @@ spec:
     replicas: 2
+  featureGates:
+    deployTektonTaskResources: true
+  tektonPipelines:
```
Sorry, I missed it. Thanks!

codingben commented on automation/common/versions.sh:
```diff
@@ -45,15 +45,18 @@ function latest_version() {
 
   # The API call sorts releases by creation timestamp, so it is enough to request only a few latest ones.
-  curl --fail -s "https://api.github.com/repos/kubevirt/${repo}/releases" | \
+  curl --fail -s "https://api.github.com/repos/${org}/${repo}/releases" | \
```
Nope, it's defined here [1] similar to `${repo}`.

[1] https://github.com/kubevirt/ssp-operator/pull/532/files#diff-ee9cd72a4fb80e8037e59a7991f40ad57757717388cf9cb8c7ad350535ccc303R56

codingben commented on README.md:
```diff
@@ -28,13 +29,17 @@ The following resource types and CRDs are needed by `ssp-operator` when deployed
 | `template.openshift.io/v1`                                              | `common-templates` operand (Kind `Template`)                              |
 | `virtualmachineclusterinstancetypes.instancetypes.kubevirt.io/v1alpha2` | `common-instancetypes` operand (Kind `VirtualMachineClusterInstancetype`) |
 | `virtualmachineclusterpreferences.instancetypes.kubevirt.io/v1alpha2`   | `common-instancetypes` operand (Kind `VirtualMachineClusterPreference`)   |
+| `pipelines.tekton.dev`                                                  | `tekton-pipelines` operand (Kind `SSP`)                                   |
```
You're right, sorry. Thanks.

codingben commented on controllers/ssp_controller.go:
```diff
@@ -465,6 +487,18 @@ func (r *sspReconciler) reconcileOperands(sspRequest *common.Request) ([]common.
 	for _, operand := range r.operands {
 		sspRequest.Logger.V(1).Info(fmt.Sprintf("Reconciling operand: %s", operand.Name()))
+
+		if operand.Name() == "tekton-pipelines" {
+			if sspRequest.Instance.Spec.FeatureGates.DeployTektonTaskResources {
+				if isMissingCRDs(sspRequest.CrdWatch, operand.RequiredCrds()) {
+					continue
+				}
+			} else {
+				sspRequest.Logger.V(1).Info("Tekton resources were not deployed, because spec.featureGates.deployTektonTaskResources is set to false")
+				continue
+			}
+		}
```
Yes, I'll try. Also I saw that I missed that it should be: `if operand.Name() == "tekton-pipelines" || operand.Name() == "tekton-tasks"`.

codingben replied:
Done.

ksimon1 commented on api/v1beta1/ssp_types.go:
```diff
@@ -91,6 +91,22 @@ type SSPSpec struct {
+	TektonPipelines *TektonPipelines `json:"tektonPipelines,omitempty"`
+
+	// FeatureGates is the configuration of the tekton operands
+	FeatureGates FeatureGates `json:"featureGates,omitempty"`
```
The feature gate should remain, as we agreed to GA this functionality and remove feature gate in 4.15. Deploying tekton via pointer is not intuitive and can lead to problems. Because even if user does not define the structures it can mean, that user does not want to set anything and would like to use default settings (deploy in the same namespace as CR)

ksimon1 commented on config/rbac/role.yaml:
```diff
@@ -110,6 +137,7 @@ rules:
   resources:
   - datavolumes
   verbs:
+  - '*'
```
the `*` should not be here

codingben replied:
@ksimon1 Can we remove it in another PR?

ksimon1 replied:
yes

codingben replied:
@ksimon1 I see that it caused by:
```
// +kubebuilder:rbac:groups=cdi.kubevirt.io,resources=datavolumes,verbs=*

```
Do you know what verbs we should specify then?

ksimon1 replied:
Everything is taken from https://github.com/kubevirt/kubevirt-tekton-tasks/tree/main/tasks

codingben replied:
Thanks. Verbs will be tracked in [1].

[1] https://issues.redhat.com/browse/CNV-31843

ksimon1 commented:
@codingben when trying to enable tekton featureGate, there is an error in log:
```
{"level":"info","ts":"2023-04-24T07:09:00Z","logger":"controllers.SSP","msg":"Resource create/update failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp":"kubevirt/ssp-sample"}
{"level":"info","ts":"2023-04-24T07:09:00Z","logger":"controllers.SSP","msg":"Operand reconciliation failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp":"kubevirt/ssp-sample"}
{"level":"error","ts":"2023-04-24T07:09:00Z","msg":"Reconciler error","controller":"ssp","controllerGroup":"ssp.kubevirt.io","controllerKind":"SSP","SSP":{"name":"ssp-sample","namespace":"kubevirt"},"namespace":"kubevirt","name":"ssp-sample","reconcileID":"998b942f-ee2c-43e6-b9c6-20b21746c6f4","error":"clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:235"}

```

ksimon1 commented on internal/common/environment.go:
```diff
@@ -16,13 +16,84 @@ import (
+	DeafultCopyTemplateIMG      = "quay.io/kubevirt/tekton-task-copy-template:" + TektonTasksVersion
+	DeafultCleanupVMIMG         = "quay.io/kubevirt/tekton-task-execute-in-vm:" + TektonTasksVersion
+	GenerateSSHKeysIMG          = "quay.io/kubevirt/tekton-task-generate-ssh-keys:" + TektonTasksVersion
+	DefaultVirtioIMG            = "quay.io/kubevirt/virtio-container-disk:v0.59.0"
```
There is a GH automation related to this variable which has to be migrated too - https://github.com/kubevirt/tekton-tasks-operator/blob/main/.github/workflows/update-virtio-version.yaml

codingben replied:
Can we do it in a new PR?

ksimon1 replied:
sure

ksimon1 commented on internal/common/labels.go:
```diff
@@ -37,7 +41,7 @@ func AddAppLabels(requestInstance *v1beta1.SSP, name string, component AppCompon
 
 	labels[AppKubernetesNameLabel] = name
 	labels[AppKubernetesComponentLabel] = component.String()
-	labels[AppKubernetesManagedByLabel] = "ssp-operator"
```
You will have to add here the `AppKubernetesManagedByValue = "tekton-tasks-operator"` variable too, because during upgrade, ssp operator will have to look for objects with `app.kubernetes.io/managed-by: tekton-tasks-operator` label and take ownership of them

codingben replied:
Okay, but don't we have only one key (`AppKubernetesManagedByValue`) to use? Can we have multiple same keys?

ksimon1 replied:
you can have only one key, but I mean it like:
Objects which already exists in cluster, have TTO's `AppKubernetesManagedByValue` label. During upgrade ssp operator will have to look for all objects with  TTO's `AppKubernetesManagedByValue` label and change to label to SSP value

ksimon1 commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+	for i := range rbs {
+		rb := &rbs[i]
+		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
+			namespace := request.Instance.Namespace
```
in the same namespace as pipelines (.Spec.Pipelines.Namespace - if possible)

codingben replied:
So we'll do: `namespace := request.Instance.Spec.TektonPipelines.Namespace`?

akrejcir replied:
yes

codingben replied:
Done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+
+var _ operands.Operand = &tektonPipelines{}
+
+func New(bundle *tektonbundle.Bundle) *tektonPipelines {
```
If we return `operands.Operand`, we no longer able to access `tektonPipelines` members that are accessed from the tests:
```
	It("New function should return object", func() {
		res := New(getMockedTestBundle())
		Expect(res.pipelines).To(HaveLen(2), "should return correct number of pipelines")
		Expect(res.configMaps).To(HaveLen(2), "should return correct number of config maps")
		Expect(res.roleBindings).To(HaveLen(2), "should return correct number of rolebindings")
	})
```
Should we then remove that test, or leave it as it is?

akrejcir replied:
We should remove the test.
In general, unit tests should test exported functionality. So functions and behaviors that are visible from outside.
Not internal data representation, which is checked by this test.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,258 @@
+)
+
+// +kubebuilder:rbac:groups=tekton.dev,resources=pipelines,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=*,resources=configmaps,verbs=list;watch;create;delete
```
Can it introduce new issues by doing that? @ksimon1 Was there a reason to use `*`?

[1] https://github.com/kubevirt/tekton-tasks-operator/blob/main/pkg/tekton-pipelines/tekton-pipelines.go#L21

ksimon1 replied:
I think I had a problem how to keep the group empty and the only way how to solve it was to use `*`

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+	})
+
+	It("RequiredCrds function should return required crds", func() {
+		tp := getMockedTektonPipelinesOperand()
```
Yep, done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+		tp := getMockedTektonPipelinesOperand()
+		crds := tp.RequiredCrds()
+
+		Expect(len(crds) > 0).To(BeTrue(), "should return required crds")
```
Done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+	}
+}
+
+func getMockedTektonPipelinesOperand() *tektonPipelines {
```
Done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+		Expect(len(functions)).To(Equal(6), "should return correct number of reconcile functions")
+	})
+
+	It("RequiredCrds function should return required crds", func() {
```
So we should remove that test?

akrejcir replied:
No, we can keep it, because it's already there.

codingben replied:
We no longer have RequiredCrds function.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+	var tp *tektonPipelines
+	var mockedRequest *common.Request
+	BeforeEach(func() {
+		tp = getMockedTektonPipelinesOperand()
```
Yep, done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -76,8 +76,8 @@ func TestTektonPipelines(t *testing.T) {
 func getMockedRequest() *common.Request {
-	log := logf.Log.WithName("tekton-pipelines-operand")
-	client := fake.NewFakeClientWithScheme(common.Scheme)
+	var log = logf.Log.WithName("tekton-pipelines-operand")
```
Okay, I've used then:
```
log := logf.Log.WithName("tekton-pipelines-operand")
```

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+	return funcs
+}
+
+func reconcileRoleBindingsFuncs(r *common.Request, rbs []rbac.RoleBinding) []common.ReconcileFunc {
```
Done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+	return funcs
+}
+
+func reconcileClusterRolesFuncs(r *common.Request, crs []rbac.ClusterRole) []common.ReconcileFunc {
```
Done.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+
+func (t *tektonPipelines) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
+	var results []common.ReconcileResult
+	var reconcileFunc []common.ReconcileFunc
```
Okay, it requires some code refactoring then, in the functions themselves.

akrejcir replied:
I didn't realize that the functions need to be changed. You can keep it this way.

akrejcir commented:
> @akrejcir Any chance, we can keep style comments to be resolved in another PR?
> 
> I'd prefer keep the code in this PR same as in `kubevirt/tekton-tasks-operator` to identify issues quickly as possible, since I know the code there works.

Ok. GitHub doesn't allow commenting on code outside of PR's, so I will add my comments here.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+	return requiredCRDs
+}
+
+func (t *tektonTasks) filterUnusedObjects() {
```
`code style` - This does not need to be a method. It can the directly inlined in `New()` function, and the slices can be filtered before the struct is created.

codingben replied:
Thanks! It solved an issue that there was no a way to test resource creation. This `filterUnusedObjects` modified only private struct variables and not the bundle (that used in unit tests).

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+}
+
+func (t *tektonTasks) filterUnusedObjects() {
+	newTasks := []pipeline.Task{}
+	for _, task := range t.tasks {
+		if _, ok := AllowedTasks[task.Name]; ok {
+			newTasks = append(newTasks, task)
+		}
+	}
```
`code style` - Can you extract this loop to a function and then call it here 4 times?

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+			request.Logger.Info(fmt.Sprintf("Changes reverted in tekton tasks: %s", r.Resource.GetName()))
+		}
+	}
+	return append(results, reconcileTektonBundleResults...), nil
```
`codes style` - It's unnecessary to use `append()`, you can return `reconcileTektonBundleResults` directly.

codingben replied:
You're right, done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+func (t *tektonTasks) Cleanup(request *common.Request) ([]common.CleanupResult, error) {
+	var objects []client.Object
+
+	for _, t := range t.tasks {
+		o := t.DeepCopy()
+		objects = append(objects, o)
+	}
```
`code style` - Can you extract this loop to a function?

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+	return common.DeleteAll(request, objects...)
+}
+
+func listDeprecatedClusterTasks(request *common.Request) ([]pipeline.ClusterTask, error) {
```
@ksimon1, This function gets all ClusterTasks with a specific label. Why are they considered deprecated?

ksimon1 replied:
The clusterTask object is deprecated https://tekton.dev/docs/pipelines/tasks/#task-vs-clustertask

akrejcir replied:
Ok. @codingben, can you please add a comment in code explaining that `ClusterTask` resource is deprecated?

akrejcir replied:
Yes, inside it, or as a docstring above the function.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+	for i := range sas {
+		sa := &sas[i]
+		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
+			namespace := request.Instance.Namespace
```
This is incorrect namespace.

codingben replied:
It should be `namespace := request.Instance.Spec.TektonPipelines.Namespace`, right?

codingben replied:
Okay, I'll do something similar as was done in another function:
```
namespace = request.Instance.Namespace
if request.Instance.Spec.TektonTasks != nil && request.Instance.Spec.TektonTasks.Namespace != "" {
		namespace = request.Instance.Spec.TektonTasks.Namespace
}
```

codingben replied:
Done.

akrejcir replied:
Please do this change.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+	for i := range rbs {
+		rb := &rbs[i]
+		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
+			namespace := request.Instance.Namespace
```
Incorrect namespace.

codingben replied:
It should be `namespace := request.Instance.Spec.TektonTasks.Namespace`, right?

codingben replied:
Done.

akrejcir replied:
Please do this change.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks_test.go:
```diff
@@ -0,0 +1,412 @@
+	var tt *tektonTasks
+	var mockedRequest *common.Request
+	BeforeEach(func() {
+		tt = getMockedTektonTasksOperand()
```
Use `New()` to create the operand.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks_test.go:
```diff
@@ -0,0 +1,412 @@
+		mockedRequest = getMockedRequest()
+	})
+
+	It("New function should return object with correct tasks", func() {
```
Can you write a similar test to check that only allowed resources are deployed, but without accessing internal struct fields of the operand?

codingben replied:
I've added similar tests to other operands:

1. `It("Should create tekton-tasks resources", func() { ... })`
2. `It("should remove tekton-tasks resources on cleanup", func() { ... })`

Done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks_test.go:
```diff
@@ -0,0 +1,412 @@
+
+})
+
+func TestTektonBundle(t *testing.T) {
```
`code style` - Can you rename the function? This is not testing `TektonBundle`.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks_test.go:
```diff
@@ -0,0 +1,412 @@
+			Expect(found).To(BeTrue(), "should return correct required crd")
+		}
+	})
+
```
Can you add more unit tests that check reconciliation? Similarly to other operands.

codingben replied:
Can we have a ticket for it, to do it later? I'd prefer to have it in a new PR.

akrejcir replied:
Ok

codingben replied:
I've added similar tests to other operands:

1. `It("Should create tekton-tasks resources", func() { ... })`
2. `It("should remove tekton-tasks resources on cleanup", func() { ... })`

Done.

akrejcir commented on tests/tests_suite_test.go:
```diff
@@ -128,6 +128,9 @@ func (s *newSspStrategy) Init() {
 				Namespace: s.GetNamespace(),
 			},
+			TektonTasks: &sspv1beta1.TektonTasks{
+				Namespace: s.GetNamespace(),
```
Can you use a different namespace here, so we can test that deployment to other namespaces works.

codingben replied:
We'll set a different namespace only to `TektonTasks`? or `TektonPipelines` as well?

To set a different namespace here, will we create `func (s *newSspStrategy) GetTektonTasksNamespace() string` and `func (s *newSspStrategy) GetTektonPipelinesNamespace() string`) if needed?

akrejcir replied:
@ksimon1 do the tasks and pipelines have to be in the same namespace?

ksimon1 replied:
they don't have to be, but if they are, user has to create clusterResolver object

codingben replied:
Done.

akrejcir replied:
@ksimon1 , I don't understand. The `ClusterResolver` needs to be created only if pipelines and tasks are in the same namespace? So if they are in different namespaces, the `ClusterResolver` is not needed?

I think I leave it to you to say in what namespace the `TektonTasks` should be deployed for functional tests.

akrejcir reviewed (commented):
I've checked commits 3 and 4, and added comments.

codingben commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -0,0 +1,179 @@
+		return nil, err
+	}
+
+	path := getPipelineBundlePath(isOpenshift)
```
Done.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,164 @@
+	)
+
+	BeforeEach(func() {
+		expectedLabels := expectedLabelsFor("tekton-pipelines", "tektonPipelines")
```
Okay, I've used `common.AppComponentTektonPipelines`.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,164 @@
+		DescribeTable("not created resource", func(res *testResource) {
+			resource := res.NewResource()
+			err := apiClient.Get(ctx, res.GetKey(), resource)
+			Expect(err).To(HaveOccurred())
```
Yep, done. Thanks!

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,164 @@
+
+		Context("with pause", func() {
+			BeforeEach(func() {
+				strategy.SkipSspUpdateTestsIfNeeded()
```
You're right, I've removed it.

codingben commented on internal/common/environment.go:
```diff
@@ -18,13 +18,77 @@ import (
 const (
 	OperatorVersionKey        = "OPERATOR_VERSION"
 	TemplateValidatorImageKey = "VALIDATOR_IMAGE"
+	OperatorNamespaceKey      = "OPERATOR_NAMESPACE"
```
Yep, removed it.

codingben commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,313 @@
+
+var _ operands.Operand = &tektonTasks{}
+
+func New(bundle *tektonbundle.Bundle) *tektonTasks {
```
Done.

codingben commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,106 @@
+		DescribeTable("not created resource", func(res *testResource) {
+			resource := res.NewResource()
+			err := apiClient.Get(ctx, res.GetKey(), resource)
+			Expect(err).To(HaveOccurred())
```
Done.

codingben commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,106 @@
-				return true
-			}, 60*time.Second, 2*time.Second).Should(BeTrue(), "conditions should have correct type, status and message")
-		})
+	Context("resource deletion", func() {
```
Yep, done. Thanks!

codingben commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,106 @@
-				Expect(sa.Labels[common.AppKubernetesManagedByLabel]).To(Equal(common.AppKubernetesManagedByValue), "managed by label should equal")
-			}
-		})
+	Context("resource creation when DeployTektonTaskResources is set to true", func() {
```
Okay, I'll also do it in TektonPipelines tests.

ksimon1 commented on tests/tests_suite_test.go:
```diff
@@ -124,6 +125,12 @@ func (s *newSspStrategy) Init() {
 				Namespace: s.GetTemplatesNamespace(),
 			},
+			TektonPipelines: &sspv1beta1.TektonPipelines{
+				Namespace: s.GetNamespace(),
```
no, the regex is there only because of one serviceAccount. Tekton does not deploy SA with name pipeline in openshift|kube namespaces. This SA is needed for running pipelines from UI. So pipelines can be deployed to any namespaces, but only when deployed to openshift|kube namespace, operator has to dpeloy extra SA.

akrejcir replied:
Ok, thanks for the explanation.
@codingben , can you use a namespace that passes the regex, so that functests can check that the extra SA is deployed correctly?

codingben replied:
Yes, very nice @akrejcir that you spotted it. Thanks.

akrejcir commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -58,6 +77,22 @@ func getPipelineBundlePath(isOpenshift bool) string {
+	return filepath.Join(tektonTasksKubernetesBundleDir, "kubevirt-tekton-tasks-kubernetes-"+common.TektonTasksVersion+".yaml")
+}
+
+func readFile(fileName string) ([][]byte, error) {
```
`code style` - Can you remove this function and just use `os.ReadFile()` instead?
You can change the signature of `decodeObjectsFromFiles()` to take variadic arguments `...[]byte`

codingben replied:
Thanks! done.

akrejcir reviewed (commented):
I've reviewed the latest changes, and I have no new comments.
I've only added a comment that I forgot to add last time.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -0,0 +1,178 @@
+			Expect(found).To(BeTrue(), "should return correct required crd")
+		}
+	})
+
```
Done.

codingben commented:
/retest

codingben commented on controllers/ssp_controller.go:
```diff
@@ -204,7 +208,7 @@ func (r *sspReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ct
 	}
 
-	if r.areCrdsMissing {
+	if r.areCrdsMissing && isTektonEnabled(r, sspRequest) {
```
Fixed. Thanks!

codingben commented on go.mod:
```diff
@@ -37,29 +37,44 @@ require (
 )
 
 require (
+	github.com/tektoncd/pipeline v0.46.0
```
Done. Thanks!

akrejcir replied:
Can you do this change in the 2nd commit `fix: tekton-pipelines operand compilation`?
I see you've done it in the 4th commit, but the 2nd commit has still version `v0.46.0`.

codingben replied:
Done.

codingben commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return nil, err
+	}
+	filesBytes := make([][]byte, len(files))
```
Done.

codingben commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+func getObject(obj map[string]interface{}, newObj interface{}) error {
+	o, err := yamlv2.Marshal(&obj)
```
Done.

codingben commented:
@ksimon1 @akrejcir @0xFelix What do you think can be the issue of the failing tests? To me it looks like some infrastructure issue. Locally `ssp-operator` deployed without issues.

akrejcir commented:
Yes, infra issue.

/retest

codingben commented:
Seems like all jobs failed because of similar error:
```
ERRO[2023-05-11T10:04:34Z] Some steps failed:                           
ERRO[2023-05-11T10:04:34Z] 
  * could not run steps: step e2e-single-node-functests failed: "e2e-single-node-functests" test steps failed: "e2e-single-node-functests" pod "e2e-single-node-functests-e2e-single-node-functests" failed: could not watch pod: the pod ci-op-8p9vv8br/e2e-single-node-functests-e2e-single-node-functests failed after 9s (failed containers: test): ContainerFailed one or more containers exited
```

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -78,76 +76,74 @@ type tektonTasks struct {
-	tt := &tektonTasks{
+func New(bundle *tektonbundle.Bundle) operands.Operand {
+	newTasks := []pipeline.Task{}
+	for _, task := range bundle.Tasks {
```
**future PR** - code style - extract this loop to a function, and then call it here 4 times.

codingben replied:
Hi, do you mean to:
```
func New(bundle *tektonbundle.Bundle) operands.Operand {
	bundle.Tasks = filterTasks(bundle.Tasks)
	bundle.ServiceAccounts = filterServiceAccounts(bundle.ServiceAccounts)
	bundle.RoleBindings = filterRoleBindings(bundle.RoleBindings)
	bundle.ClusterRoles = filterClusterRoles(bundle.ClusterRoles)

	return &tektonTasks{
		tasks:           bundle.Tasks,
		serviceAccounts: bundle.ServiceAccounts,
		roleBindings:    bundle.RoleBindings,
		clusterRoles:    bundle.ClusterRoles,
	}
}
```
Right?

codingben replied:
Woah, thank you. I tried to do it via generics, something like this:
```
func filterObjects[T metav1.Object](slice []T, trimSuffix bool) []T {
	var items []T
	for _, item := range slice {
		name := item.GetName()
		if trimSuffix {
			name = strings.TrimSuffix(name, "-task")
		}
		if _, ok := AllowedTasks[name]; ok {
			items = append(items, item)
		}
	}
	return items
}
```
It didn't work and complained with an error:
```
"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1".Task does not implement "k8s.io/apimachinery/pkg/apis/meta/v1".Object (method GetAnnotations has pointer receiver)
```

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -78,76 +76,74 @@ type tektonTasks struct {
+			newTasks = append(newTasks, task)
+		}
+	}
+	bundle.Tasks = newTasks
```
**future PR** - IMHO It's not a good design to modify the `budnle` parameter.
I would not expect a function called `New()` to modify any of it's parameters.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -78,76 +76,74 @@ type tektonTasks struct {
+	}
+	bundle.ClusterRoles = newClusterRole
+
+	return &tektonTasks{
```
**future PR** - Can you pass the local variables here, `newTasks`, `newServiceAccounts`, etc...

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks_test.go:
```diff
@@ -27,106 +31,190 @@ const (
 }
 
-func getMockedRequest() *common.Request {
+func getMockedRequest() common.Request {
```
Can you return pointer instead of a copy?

codingben replied:
Done.

akrejcir commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,113 @@
+
+	BeforeEach(func() {
+		expectedLabels := expectedLabelsFor("tekton-tasks", "tektonTasks")
+		taskResource = testResource{
```
Missing `Namespace` field.

codingben replied:
I've added `Namespace: strategy.GetNamespace()`.

akrejcir commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,113 @@
+			Resource:       &rbac.ClusterRole{},
+			ExpectedLabels: expectedLabels,
+		}
+		roleBindingResource = testResource{
```
Missing `Namespace` field.

codingben replied:
I've added `Namespace: strategy.GetNamespace()`.

akrejcir commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,113 @@
+			Resource:       &rbac.RoleBinding{},
+			ExpectedLabels: expectedLabels,
+		}
+		serviceAccountResource = testResource{
```
Missing `Namespace` field.

codingben replied:
I've added `Namespace: strategy.GetNamespace()`.

akrejcir commented on tests/tekton-tasks_test.go:
```diff
@@ -1,362 +1,113 @@
-		})
-	})
-	Context("multiple CRs deployed", func() {
+	Context("resource creation when DeployTektonTaskResources is set to false", func() {
```
This context can be removed, because SSP operator does not touch tekton tasks if the feature gate is false.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -45,46 +42,49 @@ type tektonPipelines struct {
-		&rbac.ClusterRole{},
+func (t *tektonPipelines) WatchClusterTypes() []operands.WatchType {
+	return []operands.WatchType{
+		{Object: &pipeline.Pipeline{}, Crd: tektonCrd, WatchFullObject: true},
```
This makes the teketon CRD required. SSP operator will wait until it exists.

@ksimon1, is it acceptable?

**future PR** - SSP should not wait for tekton CRD, if the feature gate is `false`.

ksimon1 replied:
yes, that can be done follow up PR

codingben replied:
Hi, so if Tekton CRD doesn't exist, Tekton operands won't reconcile anymore.

akrejcir commented on internal/common/resource.go:
```diff
@@ -490,6 +491,9 @@ func defaultUpdateFunc(newObj, foundObj client.Object) {
 	case *promv1.ServiceMonitor:
 		foundObj.(*promv1.ServiceMonitor).Spec = newTyped.Spec
 
+	case *v1beta1.Task:
```
**future PR** - code style - please use a different name for the import. `v1beta1.Task` is unclear. `tekton.Task` would be more clear.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -148,7 +148,7 @@ func reconcileTektonPipelinesFuncs(pipelines []pipeline.Pipeline) []common.Recon
 					for i, param := range foundPipeline.Spec.Params {
 						if strings.HasPrefix(param.Name, "virtioContainer") {
-							foundPipeline.Spec.Params[i].Default = &pipeline.ParamValue{
+							foundPipeline.Spec.Params[i].Default = &pipeline.ArrayOrString{
```
This change is not needed.

codingben replied:
Done.

akrejcir reviewed (commented):
I've added more comments.

The ones marked with **future PR** or "code style" can be done in another PR.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines_test.go:
```diff
@@ -26,59 +31,171 @@ const (
 }
 
-func getMockedRequest() *common.Request {
+func getMockedRequest() common.Request {
```
Done.

codingben commented on tests/tests_suite_test.go:
```diff
@@ -175,6 +183,11 @@ func (s *newSspStrategy) GetVmConsoleProxyNamespace() string {
 }
 
+func (s *newSspStrategy) GetTektonPipelinesNamespace() string {
+	const tektonPipelinesNamespace = "openshift-functests"
```
Done.

codingben commented on tests/tests_suite_test.go:
```diff
@@ -319,6 +332,13 @@ func (s *existingSspStrategy) GetVmConsoleProxyNamespace() string {
+	if s.ssp == nil || s.ssp.Spec.TektonPipelines == nil {
+		panic("Strategy is not initialized")
+	}
+	return s.ssp.Spec.TektonPipelines.Namespace
```
So we'll check to `.ssp.Spec.TektonPipelines.Namespace` as well, and return some default value instead of Panic?

akrejcir replied:
That depends on Karel's answer to: https://github.com/kubevirt/ssp-operator/pull/532#discussion_r1190924235

If we support empty `.ssp.Spec.TektonPipelines.Namespace`, then there should be the same default as in production code. And if we don't support `""`, then it can panic.

codingben commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -45,46 +42,49 @@ type tektonPipelines struct {
-
 func (t *tektonPipelines) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
-	var results []common.ReconcileResult
+	if request.Instance.Spec.FeatureGates.DeployTektonTaskResources {
```
Done.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,168 @@
+
+	BeforeEach(func() {
+		expectedLabels := expectedLabelsFor("tekton-pipelines", common.AppComponentTektonPipelines)
+		pipelineResource = testResource{
```
I've added `Namespace: strategy.GetTektonPipelinesNamespace()`.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,168 @@
+			Resource:       &rbac.ClusterRole{},
+			ExpectedLabels: expectedLabels,
+		}
+		roleBindingResource = testResource{
```
I've added `Namespace: strategy.GetTektonPipelinesNamespace()`.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,168 @@
+			Resource:       &rbac.RoleBinding{},
+			ExpectedLabels: expectedLabels,
+		}
+		configMapResource = testResource{
```
I've added `Namespace: strategy.GetTektonPipelinesNamespace()`.

codingben commented on tests/tekton-pipelines_test.go:
```diff
@@ -1,162 +1,168 @@
+					reflect.DeepEqual(old.BinaryData, new.BinaryData)
+			},
+		}
+		serviceAccountResource = testResource{
```
I've added `Namespace: strategy.GetTektonPipelinesNamespace()`.

codingben commented on go.mod:
```diff
@@ -37,29 +38,39 @@ require (
 )
 
 require (
+	contrib.go.opencensus.io/exporter/ocagent v0.7.1-0.20200907061046-05415f1de66d // indirect
```
Done.

codingben commented:
/retest

ksimon1 commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -128,16 +128,17 @@ func (t *tektonPipelines) Cleanup(request *common.Request) ([]common.CleanupResu
 		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
-			namespace := request.Instance.Spec.Pipelines.Namespace
-			p.Namespace = namespace
+			if request.Instance.Spec.TektonPipelines != nil && request.Instance.Spec.TektonPipelines.Namespace != "" {
```
in the same namespace as operator

codingben replied:
So we'll add `p.Namespace = "kubevirt"` before the `if` condition?

akrejcir replied:
@ksimon1 the **operator** or the SSP **CR**?

akrejcir replied:
@codingben , hardcoding namespace is not a good design. There is a function to get the operator namespace: `common.GetOperatorNamespace()`

ksimon1 replied:
You are right, operator can be deployed in a different namespace than CR. I would vote for CR namespace, because the same way it is done for common templates

ksimon1 commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -165,16 +166,12 @@ func reconcileConfigMapsFuncs(configMaps []v1.ConfigMap) []common.ReconcileFunc
 		funcs = append(funcs, func(request *common.Request) (common.ReconcileResult, error) {
-			namespace := request.Instance.Spec.Pipelines.Namespace
-			cm.Namespace = namespace
+			if request.Instance.Spec.TektonPipelines != nil && request.Instance.Spec.TektonPipelines.Namespace != "" {
```
in the same namespace as operator

codingben commented:
@akrejcir @ksimon1 Hi, currently functional tests are failing locally. I've pushed [1] to add/change couple of things. The failure reason that I'm trying to investigate is:
```
│ {"level":"info","ts":"2023-05-15T11:00:13Z","logger":"controllers.SSP","msg":"Resource create/update failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp │
│ {"level":"info","ts":"2023-05-15T11:00:13Z","logger":"controllers.SSP","msg":"Operand reconciliation failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp │
│ {"level":"error","ts":"2023-05-15T11:00:13Z","msg":"Reconciler error","controller":"ssp","controllerGroup":"ssp.kubevirt.io","controllerKind":"SSP","SSP":{"name":"test-ssp","namespace":"ssp-o │  
```

[1] https://github.com/kubevirt/ssp-operator/compare/8f1b0e866be11ee9211e78f0b5135764189729b3..18043bd29490246bc068d2fc6b4793c0cdfd0839

ksimon1 commented:
> @akrejcir @ksimon1 Hi, currently functional tests are failing locally. I've pushed [1] to add/change couple of things. The failure reason that I'm trying to investigate is:
> 
> ```
> │ {"level":"info","ts":"2023-05-15T11:00:13Z","logger":"controllers.SSP","msg":"Resource create/update failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp │
> │ {"level":"info","ts":"2023-05-15T11:00:13Z","logger":"controllers.SSP","msg":"Operand reconciliation failed: clusterroles.rbac.authorization.k8s.io \"modify-data-object-task\" not found","ssp │
> │ {"level":"error","ts":"2023-05-15T11:00:13Z","msg":"Reconciler error","controller":"ssp","controllerGroup":"ssp.kubevirt.io","controllerKind":"SSP","SSP":{"name":"test-ssp","namespace":"ssp-o │  
> ```
> 
> [1] https://github.com/kubevirt/ssp-operator/compare/8f1b0e866be11ee9211e78f0b5135764189729b3..18043bd29490246bc068d2fc6b4793c0cdfd0839

Switch these two lines: https://github.com/kubevirt/ssp-operator/pull/532/files#diff-32a274b974ae70f092b3f061d72858382354b8f710d468d80883475162e3920aR87 (edited) 
So it will look like:
```
data_sources.New(templatesBundle.DataSources),
tektonTasksOperand,
tektonPipelinesOperand,
```
and it will work

codingben commented:
@ksimon1 Thanks! I changed it, let's see if it works now.

codingben commented:
Okay so now we know why functional tests are failing, thanks to @akrejcir that found it, it seems like the functional tests configured by name that aren't the same as in the bundle files. I'll fix it tomorrow.

codingben commented:
@akrejcir @ksimon1 I think we're missing something here [1]. Do we need to update it to create a new namespace, `openshift-pipelines-ssp-functests`?

In functional tests, we have deployment in:
- Tekton Tasks in `ssp-operator-functests`
- Tekton Pipelines in `openshift-pipelines-ssp-functests`

[1] https://github.com/kubevirt/ssp-operator/blob/master/automation/e2e-upgrade-functests/run.sh#L40-L41

codingben commented:
/test e2e-upgrade-functests

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -45,46 +42,48 @@ type tektonPipelines struct {
-
 func (t *tektonPipelines) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
-	var results []common.ReconcileResult
+	if !request.Instance.Spec.FeatureGates.DeployTektonTaskResources {
```
`FeatureGates` pointer can be nil.

codingben replied:
Done. I've added `request.Instance.Spec.FeatureGates != nil`.

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -0,0 +1,260 @@
+
+func (t *tektonPipelines) Cleanup(request *common.Request) ([]common.CleanupResult, error) {
+	var objects []client.Object
+	for _, p := range t.pipelines {
```
These objects need to have the correct `namespace` before deleting them. Otherwise the operator may try to delete objects from a wrong namespace.

codingben replied:
Thanks!

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -163,96 +163,76 @@ func reconcileTektonPipelinesFuncs(pipelines []pipeline.Pipeline) []common.Recon
 			namespace := request.Instance.Namespace
-			if sa.Namespace == "" {
-				sa.Namespace = namespace
+			if serviceAccount.Namespace == "" {
```
IMHO, it's not correct to use the namespace stored in the bundle. I think we should overwrite the namespace with the one specified in `.Spec.TektonPipelines.Namespace`.

@ksimon1 What do you think?

ksimon1 replied:
all objects related to pipelines (service accounts, role, role binding) should be deployed to the same namespace as pipelines (in this case value from .Spec.TektonPipelines.Namespace - if possible)

codingben replied:
Okay, so instead of:
```
namespace := request.Instance.Namespace
if serviceAccount.Namespace == "" {
	serviceAccount.Namespace = namespace
}
```
We'll do:
```
if request.Instance.Spec.TektonPipelines != nil && request.Instance.Spec.TektonPipelines.Namespace != "" {
	serviceAccount.Namespace = request.Instance.Spec.TektonPipelines.Namespace
}
```
In all resources?

akrejcir replied:
yes, but for Tekton Tasks, it will be the other field: `request.Instance.Spec.TektonTasks.Namespace`.

And when the namespace in `request.Instance.Spec.TektonPipelines.Namespace` is `""`, then the namepsace on resource needs to be set to the `request.Namespace`.

codingben replied:
Yes, of course, in Tekton Tasks operand it's different. So then:
```
serviceAccount.namespace := request.Instance.Namespace
if request.Instance.Spec.TektonPipelines != nil && request.Instance.Spec.TektonPipelines.Namespace != "" {
	serviceAccount.Namespace = request.Instance.Spec.TektonPipelines.Namespace
}
```

ksimon1 replied:
yes and the same should be done for all objects (except ClusterRole)

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-pipelines/tekton-pipelines.go:
```diff
@@ -163,96 +163,76 @@ func reconcileTektonPipelinesFuncs(pipelines []pipeline.Pipeline) []common.Recon
 			namespace := request.Instance.Namespace
-			if rb.Namespace == "" {
-				rb.Namespace = namespace
+			if roleBinding.Namespace == "" {
```
Same question as above.

akrejcir commented on tests/tekton-pipelines_test.go:
```diff
@@ -6,157 +6,258 @@ import (
 	rbac "k8s.io/api/rbac/v1"
+	ssp "kubevirt.io/ssp-operator/api/v1beta1"
+	"kubevirt.io/ssp-operator/internal/common"
+	tekton_tasks "kubevirt.io/ssp-operator/internal/operands/tekton-tasks"
```
`tekton_tasks` package does not exist yet in this commit. Please remove the import, so it can compile.

codingben replied:
That's really weird that it exists here. I don't know how I missed it.

akrejcir commented on tests/tekton-pipelines_test.go:
```diff
@@ -6,157 +6,258 @@ import (
+		strategy.SkipSspUpdateTestsIfNeeded()
+
+		updateSsp(func(foundSsp *ssp.SSP) {
+			foundSsp.Spec.FeatureGates.DeployTektonTaskResources = false
```
`FeatureGates` may be `nil`

codingben replied:
Okay, in that case we then simply just do:
```
if foundSsp.Spec.FeatureGates != nil {
   foundSsp.Spec.FeatureGates.DeployTektonTaskResources = false
}
```
Right?

akrejcir replied:
yes.

akrejcir replied:
But also check for `nil` at the place that sets it to `true`. I forgot to comment there.

codingben replied:
Yes, I'll add nil check everywhere in the code where we use `foundSsp.Spec.FeatureGates.DeployTektonTaskResources`.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -78,76 +76,73 @@ type tektonTasks struct {
-			newSA = append(newSA, sa)
-		}
+func (t *tektonTasks) Reconcile(request *common.Request) ([]common.ReconcileResult, error) {
+	if !request.Instance.Spec.FeatureGates.DeployTektonTaskResources {
```
`FeatureGates` may be nil.

codingben replied:
Done.

akrejcir commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -165,12 +160,11 @@ func (t *tektonTasks) Reconcile(request *common.Request) ([]common.ReconcileResu
 
 func (t *tektonTasks) Cleanup(request *common.Request) ([]common.CleanupResult, error) {
 	var objects []client.Object
-
```
The resources need correct namespace.

codingben replied:
Thanks!

codingben commented:
/test e2e-functests

codingben commented:
@akrejcir @ksimon1 All tests passed, your `/approve` is very much appreciated :)

ksimon1 commented on docs/development.md:
```diff
@@ -157,3 +157,11 @@ for more docs.
+
+#### `tekton-pipelines` operand
+
+Installs the Tekton Pipelines found in `data/tekton-pipelines`. A PR for updating the bundle is automatically created when a new version of the [kubevirt-tekton-tasks](https://github.com/kubevirt/kubevirt-tekton-tasks) is released.
```
There is no PR for updating pipelines. The PR is only for tekton tasks.
Can be fixed in a follow up PRs

codingben replied:
Ah okay. The pipelines bundle yaml files will be updated manually if needed?

ksimon1 replied:
yes

ksimon1 commented on internal/common/resource.go:
```diff
@@ -490,6 +491,9 @@ func defaultUpdateFunc(newObj, foundObj client.Object) {
 	case *promv1.ServiceMonitor:
 		foundObj.(*promv1.ServiceMonitor).Spec = newTyped.Spec
 
+	case *tekton.Task:
```
Pipelines should be here too?

codingben replied:
I didn't include it because if it has some uncommon logic, so it would be better to keep it as it is.

ksimon1 replied:
ok

ksimon1 commented:
/lgtm
The comments I posted can be fixed in a follow up PRs

akrejcir commented:
/approve

codingben commented on internal/tekton-bundle/tekton-bundle_test.go:
```diff
@@ -0,0 +1,49 @@
+		pipelineFiles, err := readFolder(pipelinePath)
+		Expect(err).ToNot(HaveOccurred())
+		files := [][]byte{}
+		files = append(files, pipelineFiles...)
```
Hi, `files` variable is needed because the type of `taskFile` is `[]byte` and the type of `pipelineFiles` is `[][]byte` and I'm not sure how we can not use it.

akrejcir replied:
I think I posted this review comment on the commit that added `pipelineFiles`, but not `taskFile`.
Now it is irrelevant when the PR is merged.

codingben commented on internal/tekton-bundle/tekton-bundle.go:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	path := getPipelineBundlePath(isOpenshift)
+	files, err := readFolder(path)
```
Hi, how do you think we can achieve it? Generators in Go aren't supported, from what I see it can be achieved via Goroutines/Channels. This is what you mean?

akrejcir replied:
Simply merge the functions `readFolder()` and `decodeObjectsFromFiles()`. 
The outer loops in both functions can be combined, so that each file is processed after it is loaded, and then the next file is loaded and porcessed.

codingben replied:
I wish I could do it, but that `decodeObjectsFromFiles()` is used in both different functions:
```
func ReadTasksBundle(isOpenshift bool) (*Bundle, error) {
	var files [][]byte
	path := getTasksBundlePath(isOpenshift)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	files = append(files, data)

	tektonObjs, err := decodeObjectsFromFiles(files)
	if err != nil {
		return nil, err
	}

	return tektonObjs, nil
}

func ReadPipelineBundle() (*Bundle, error) {
	path := getPipelineBundlePath()
	files, err := readFolder(path)
	if err != nil {
		return nil, err
	}

	tektonObjs, err := decodeObjectsFromFiles(files)
	if err != nil {
		return nil, err
	}

	return tektonObjs, nil
}
```
I'll take a look.

akrejcir replied:
You could, for example, change the argument of `decodeObjectsFromFiles()` function to be a list of filenames, and then inside it would open and load the files one by one.

codingben replied:
Done.

codingben commented on internal/operands/tekton-tasks/tekton-tasks.go:
```diff
@@ -0,0 +1,305 @@
+// +kubebuilder:rbac:groups=subresources.kubevirt.io,resources=virtualmachines/restart;virtualmachines/start;virtualmachines/stop,verbs=update
+// +kubebuilder:rbac:groups=template.openshift.io,resources=templates,verbs=get;list;watch;create;patch;update;delete
+// +kubebuilder:rbac:groups=template.openshift.io,resources=processedtemplates,verbs=create
+// +kubebuilder:rbac:groups=cdi.kubevirt.io,resources=datavolumes,verbs=*
+// +kubebuilder:rbac:groups=cdi.kubevirt.io,resources=datasources,verbs=get;create;delete
+// +kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachines/finalizers,verbs=*
+// +kubebuilder:rbac:groups=*,resources=persistentvolumeclaims,verbs=*
+// +kubebuilder:rbac:groups=*,resources=pods,verbs=create
+// +kubebuilder:rbac:groups=*,resources=secrets,verbs=*
```
Hi @ksimon1, how I can learn which `verbs` are needed here?
```
// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=*
// +kubebuilder:rbac:groups=core,resources=secrets,verbs=*
```
I know that the `groups` is `core`.

ksimon1 replied:
Hi, all these RBACs come from  https://github.com/kubevirt/kubevirt-tekton-tasks/tree/main/tasks, specifically from https://github.com/kubevirt/kubevirt-tekton-tasks/blob/main/tasks/create-vm-from-manifest/manifests/create-vm-from-manifest.yaml#L99 (other tasks might require these permissions too)

codingben replied:
Thanks. I've updated only the API groups, and verbs will be tracked in [1].

[1] https://issues.redhat.com/browse/CNV-31843
